dymos/examples/racecar/carODE.py

```diff
@@ -0,0 +1,262 @@
+
+        self.declare_partials(of="power", wrt="V", rows=arange, cols=arange)
+        self.declare_partials(of="power", wrt="S_rr", rows=arange, cols=arange)
+        self.declare_partials(of="power", wrt="S_rl", rows=arange, cols=arange)
+        self.declare_partials(of="power", wrt="S_fl", rows=arange, cols=arange)
+        self.declare_partials(of="power", wrt="S_fr", rows=arange, cols=arange)
+
+    def compute(self, inputs, outputs):
+        omega = inputs["omega"]
+        V = inputs["V"]
+        n = inputs["n"]
+        lamb = inputs["lambda"]
+        delta = inputs["delta"]
+        alpha = inputs["alpha"]
+        S_fl = inputs["S_fl"]
+        S_fr = inputs["S_fr"]
+        S_rl = inputs["S_rl"]
+        S_rr = inputs["S_rr"]
+        F_fl = inputs["F_fl"]
+        F_fr = inputs["F_fr"]
+        F_rl = inputs["F_rl"]
+        F_rr = inputs["F_rr"]
+        kappa = inputs["kappa"]
+        M = inputs["M"]
+        rho = inputs["rho"]
+        Iz = inputs["Iz"]
+        a = inputs["a"]
+        b = inputs["b"]
+        tw = inputs["tw"]
+        CdA = inputs["CdA"]
+
+        F_all = F_fl + F_fr + F_rl + F_rr
+        S_all = S_fl + S_fr + S_rl + S_rr
+        F_front = F_fl + F_fr
+        S_front = S_fl + S_fr
+
+        # drag equation
+        drag = 0.5 * rho * CdA * V ** 2
+
+        outputs["sdot"] = (V * np.cos(alpha - lamb)) / (1 - n * kappa)
+        outputs["ndot"] = V * np.sin(alpha - lamb)
+        outputs["alphadot"] = omega - (kappa * V * np.cos(alpha - lamb)) / (1 - n * kappa)
+        Vdot = S_all / M - delta * F_front / M - drag / M - omega * V * lamb
+        outputs["Vdot"] = Vdot
+        outputs["lambdadot"] = (
+                    omega - Vdot * lamb / V - delta * S_front / (M * V) - F_all / (M * V))
+        outputs["omegadot"] = ((a * (F_fr + F_fl)) / Iz - (b * (F_rr + F_rl)) / Iz + (
+                    tw * (-S_rr + S_rl - S_fr + S_fl)) / Iz)
+        outputs["power"] = V * S_all
+
+    def compute_partials(self, inputs, jacobian):
+        omega = inputs["omega"]
+        V = inputs["V"]
+        n = inputs["n"]
+        lamb = inputs["lambda"]
+        delta = inputs["delta"]
+        alpha = inputs["alpha"]
+        S_fl = inputs["S_fl"]
+        S_fr = inputs["S_fr"]
+        S_rl = inputs["S_rl"]
+        S_rr = inputs["S_rr"]
+        F_fl = inputs["F_fl"]
+        F_fr = inputs["F_fr"]
+        F_rl = inputs["F_rl"]
+        F_rr = inputs["F_rr"]
+        kappa = inputs["kappa"]
+        M = inputs["M"]
+        rho = inputs["rho"]
+        Iz = inputs["Iz"]
+        a = inputs["a"]
+        b = inputs["b"]
+        tw = inputs["tw"]
+        CdA = inputs["CdA"]
+
+        F_all = F_fl + F_fr + F_rl + F_rr
+        S_all = S_fl + S_fr + S_rl + S_rr
+        F_front = F_fl + F_fr
+        S_front = S_fl + S_fr
+
+        ddrag_dv = 2 * CdA * V
+
+        jacobian["sdot", "V"] = np.cos(alpha - lamb) / (1 - n * kappa)
```

Comment:
A fair amount can be done algebraically to make these calculations more efficient, keeping in mind that many of these computations are done on a vector of points around the racetrack. Most will give small gains and the code may be more readable without some of them, but the big one here is all of these calls to sin and cos.  You should calculate those once and save them because that can be slow compared to all these adds and multiplies.